README.md
# Logstash Plugin

[![Travis Build Status](https://travis-ci.org/logstash-plugins/logstash-output-zabbix.svg)](https://travis-ci.org/logstash-plugins/logstash-output-zabbix)

This is a plugin for [Logstash](https://github.com/elastic/logstash).

It is fully free and fully open source. The license is Apache 2.0, meaning you are pretty much free to use it however you want in whatever way.

## Documentation

Logstash provides infrastructure to automatically generate documentation for this plugin. We use the asciidoc format to write documentation so any comments in the source code will be first converted into asciidoc and then into html. All plugin documentation are placed under one [central location](http://www.elastic.co/guide/en/logstash/current/).

- For formatting code or config example, you can use the asciidoc `[source,ruby]` directive
- For more asciidoc formatting tips, see the excellent reference here https://github.com/elastic/docs#asciidoc-guide

## Need Help?

Need help? Try #logstash on freenode IRC or the https://discuss.elastic.co/c/logstash discussion forum.

## Developing

### 1. Plugin Developement and Testing

#### Code
- To get started, you'll need JRuby with the Bundler gem installed.

- Create a new plugin or clone and existing from the GitHub [logstash-plugins](https://github.com/logstash-plugins) organization. We also provide [example plugins](https://github.com/logstash-plugins?query=example).

- Install dependencies
```sh
bundle install
```

#### Test

- Update your dependencies

```sh
bundle install
```

- Run tests

```sh
bundle exec rspec
```

### 2. Running your unpublished Plugin in Logstash

#### 2.1 Run in a local Logstash clone

- Edit Logstash `Gemfile` and add the local plugin path, for example:
```ruby
gem "logstash-filter-awesome", :path => "/your/local/logstash-filter-awesome"
```
- Install plugin
```sh
# Logstash 2.3 and higher
bin/logstash-plugin install --no-verify

# Prior to Logstash 2.3
bin/plugin install --no-verify

```
- Run Logstash with your plugin
```sh
bin/logstash -e 'filter {awesome {}}'
```
At this point any modifications to the plugin code will be applied to this local Logstash setup. After modifying the plugin, simply rerun Logstash.

#### 2.2 Run in an installed Logstash

You can use the same **2.1** method to run your plugin in an installed Logstash by editing its `Gemfile` and pointing the `:path` to your local plugin development directory or you can build the gem and install it using:

- Build your plugin gem
```sh
gem build logstash-filter-awesome.gemspec
```
- Install the plugin from the Logstash home
```sh
# Logstash 2.3 and higher
bin/logstash-plugin install --no-verify

# Prior to Logstash 2.3
bin/plugin install --no-verify

```
- Start Logstash and proceed to test the plugin

## Contributing

All contributions are welcome: ideas, patches, documentation, bug reports, complaints, and even something you drew up on a napkin.

Programming is not a required skill. Whatever you've seen about open source and maintainers or community members  saying "send patches or die" - you will not see that here.

It is more important to the community that you are able to contribute.

For more information about contributing, see the [CONTRIBUTING](https://github.com/elastic/logstash/blob/master/CONTRIBUTING.md) file.


CHANGELOG.md
## 3.0.5
  - Docs: Set the default_codec doc attribute.

## 3.0.4
  - Update gemspec summary

## 3.0.3
  - Fix some documentation issues

## 3.0.1
  - Relax constraint on logstash-core-plugin-api to >= 1.60 <= 2.99

# 3.0.0
  - Breaking: Updated plugin to use new Java Event APIs
  - Use zabbix_protocol 0.1.5 or greater (tasty fixes for Zabbix)
  - Make TravisCI use JRuby 1.7.25, and have Travis run the unit & integration tests.
# 2.0.2
  - Depend on logstash-core-plugin-api instead of logstash-core, removing the need to mass update plugins on major releases of logstash
# 2.0.1
  - New dependency requirements for logstash-core for the 5.0 release
## 2.0.0
 - Plugins were updated to follow the new shutdown semantic, this mainly allows Logstash to instruct input plugins to terminate gracefully,
   instead of using Thread.raise on the plugins' threads. Ref: https://github.com/elastic/logstash/pull/3895
 - Dependency on logstash-core update to 2.0
 - Update to allow semicolons (or not, e.g. Zabbix 2.0). #6 (davmrtl)
 - Update to prevent Logstash from crashing when the Zabbix server becomes unavailable while the plugin is sending data. #9 (dkanbier)


spec/helpers/stevedore.rb
# Helper methods for Docker testing

module Stevedore
  require 'docker'

  # Figure out which IP address the Docker host is at
  def get_host_ip
    # Let the crazy one-liner definition begin:
    # Docker.url.split(':')[1][2..-1]
    # Docker.url = tcp://192.168.123.205:2375
    #   split(':') = ["tcp", "//192.168.123.205", "2375"]
    #   [1] = "//192.168.123.205"
    #   [2..-1] = "192.168.123.205"
    # This last bit prunes the leading //
    url = Docker.url
    case url.split(':')[0]
    when 'unix'
      ip = "127.0.0.1"
    when 'tcp'
      ip = url.split(':')[1][2..-1]
    end
    ip
  end

  # Build an image from a Dockerfile
  def build_image(dockerfile_path)
    dockerfile = IO.read(dockerfile_path)
    Docker::Image.build(dockerfile)
  end

  def delete_image(image)
    if image_exists?(image.id)
      image.remove(:force => true)
    end
  end

  # Return the image object identified by repository:tag
  def get_image(repository, tag)
    images = Docker::Image.all.select { |i|
      i.info["RepoTags"].include? "#{repository}:#{tag}"
    }
    # I don't think it's possible to have multiple images from the same
    # repository with the same tag, so this shouldn't be an issue.
    images[0]
  end

  def image_exists?(id)
    begin
      Docker::Image.get(id)
      true
    rescue Docker::Error::NotFoundError
      false
    end
  end

  # Return the randomized port number associated with the exposed port.
  def get_randomized_port(container, exposed_port, protocol="tcp")
    # Get all mapped ports
    ports = container.json["NetworkSettings"]["Ports"]
    # We're going to expect 1:1 mapping here
    ports["#{exposed_port.to_s}/#{protocol}"][0]["HostPort"].to_i
  end

  # Create a docker container from an image (or repository:tag string), name,
  # and optional extra args.
  def create_container(image, name, extra_args={})
    # If this ends up getting integrated, we'll use @logger.error here
    case image.class.to_s
    when "Docker::Image"
      i = image.id
    when "String"
      if image.include? ':' # repository:tag format
        i = image
      else
        puts "Image string must be in 'repository:tag' format."
        return
      end
    else
      puts "image must be Docker::Image or in 'repository:tag' format"
      return
    end
    Docker.options = { :write_timeout => 300, :read_timeout => 300 }
    Docker.validate_version!
    # Don't change the non-capitalized 'name' here.  The Docker API gem extracts
    # this key and uses it to name the container on create.
    main_args = {
      'name' => name,
      'Hostname' => name,
      'Image' => i,
      'PublishAllPorts' => true,
    }
    Docker::Container.create(main_args.merge(extra_args))
  end

  # Wrapper to clean up container in a single call
  # name can be a name or id
  def cleanup_container(name)
    container = Docker::Container.get(name)
    if container
      container.stop
      container.delete(:force => true)
    end
  end

  # This is useful if you're building from Dockerfiles each run
  def cleanup_image(image)
    if image_exists?(image.id)
      image.remove(:force => true)
    end
  end

end


spec/outputs/zabbix_spec.rb
require "logstash/devutils/rspec/spec_helper"
require "logstash/outputs/zabbix"
require "logstash/codecs/plain"
require "logstash/event"
require "zabbix_protocol"
require "socket"
require "timeout"
require "longshoreman"
require_relative "../helpers/zabbix_helper"

RSpec.configure do |c|
  c.include ZabbixHelper
end

NAME = "logstash-output-zabbix-#{rand(999).to_s}"
IMAGE = "untergeek/logstash_output_zabbix_rspec"
TAG = "latest"
ZABBIX_PORT = 10051

describe LogStash::Outputs::Zabbix do
  # Building block "lets"
  let(:port) { ZABBIX_PORT }
  let(:host) { "127.0.0.1" }
  let(:timeout) { 1 }
  let(:timestamp) { "2015-04-15T15:39:42Z" }
  let(:epoch) { 1429112382 } # This is the epoch value of the above timestamp
  let(:zabhost) { "zabbix.example.com" }
  let(:zabkey) { "zabbix.key" }
  let(:message) { "This is a log entry." }

  # Assembled "lets"
  let(:event_hash) {
    {
      "@timestamp" => timestamp,
      "message" => message,
      "zabhost" => zabhost,
      "zabkey"  => zabkey,
    }
  }
  let(:multi_value) {
    {
      "@timestamp" => timestamp,
      "message"    => message,
      "zabhost"    => zabhost,
      "key1"       => "multi1",
      "val1"       => "value1",
      "key2"       => "multi2",
      "val2"       => "value2",
    }
  }
  let(:zabout) {
    {
      "zabbix_server_host" => host,
      "zabbix_server_port" => port,
      "zabbix_host"        => "zabhost",
      "zabbix_key"         => "zabkey",
      "timeout"            => timeout,
    }
  }
  let(:mzabout) {
    {
      "zabbix_server_host" => host,
      "zabbix_server_port" => port,
      "zabbix_host"        => "zabhost",
      "multi_value"        => ["key1", "val1", "key2", "val2"],
      "timeout"            => timeout,
    }
  }

  # Finished object "lets"
  let(:string_event) { LogStash::Event.new(event_hash) }
  let(:output) { LogStash::Outputs::Zabbix.new(zabout) }

  before do
    output.register
  end

  describe "Unit Tests" do

    describe "#field_check" do

      context "when expected field not found" do
        subject { output.field_check(string_event, "not_appearing") }
        it "should return false" do
          expect(subject).to eq(false)
        end
      end

      context "when expected field found" do
        subject { output.field_check(string_event, "zabhost") }
        it "should return true" do
          expect(subject).to eq(true)
        end
      end

    end # "#field_check"

    describe "#format_request (single key/value pair)" do

      context "when it receives an event" do
        # {
        #   "request" => "sender data",
        #   "data" => [{
        #     "host" => "zabbix.example.com",
        #     "key" => "zabbix.key",
        #     "value" => "This is a log entry.",
        #     "clock" => 1429112382
        #   }],
        #   "clock" => 1429112394
        # }
        subject { output.format_request(string_event) }
        it "should return a Zabbix sender data object with the correct host" do
          expect(subject['data'][0]['host']).to eq(zabhost)
        end
        it "should return a Zabbix sender data object with the correct key" do
          expect(subject['data'][0]['key']).to eq(zabkey)
        end
        it "should return a Zabbix sender data object with the correct value" do
          expect(subject['data'][0]['value']).to eq(message)
        end
        it "should return a Zabbix sender data object with the correct clock from @timestamp" do
          expect(subject['data'][0]['clock']).to eq(epoch)
        end
        it "should return a Zabbix sender data object with the correct clock" do
          diff = Time.now.to_i - subject['clock']
          expect(diff).to be < 3 # It should take less than 3 seconds for this.
        end
      end

    end # "#format_request (single key/value pair)"

    describe "#format_request (multiple key/value pairs)" do
      let(:multival_event) { LogStash::Event.new(multi_value) }
      let(:m_output) { LogStash::Outputs::Zabbix.new(mzabout) }

      context "when it receives an event and is configured for multiple values" do
        subject { m_output.format_request(multival_event) }
        it "should return a Zabbix sender data object with a data array" do
          expect(subject['data'].length).to eq(2)
        end
        it "should return a Zabbix sender data object with the correct host [0]" do
          expect(subject['data'][0]['host']).to eq(zabhost)
        end
        it "should return a Zabbix sender data object with the correct key [0]" do
          expect(subject['data'][0]['key']).to eq('multi1')
        end
        it "should return a Zabbix sender data object with the correct value [0]" do
          expect(subject['data'][0]['value']).to eq('value1')
        end
        it "should return a Zabbix sender data object with the correct clock from @timestamp [0]" do
          expect(subject['data'][0]['clock']).to eq(epoch)
        end
        it "should return a Zabbix sender data object with the correct host [1]" do
          expect(subject['data'][1]['host']).to eq(zabhost)
        end
        it "should return a Zabbix sender data object with the correct key [1]" do
          expect(subject['data'][1]['key']).to eq('multi2')
        end
        it "should return a Zabbix sender data object with the correct value [1]" do
          expect(subject['data'][1]['value']).to eq('value2')
        end
        it "should return a Zabbix sender data object with the correct clock from @timestamp [1]" do
          expect(subject['data'][1]['clock']).to eq(epoch)
        end
        it "should return a Zabbix sender data object with the correct clock" do
          diff = Time.now.to_i - subject['clock']
          expect(diff).to be < 3 # It should take less than 3 seconds for this.
        end
      end

    end # "#format_request (multiple key/value pairs)"

    describe "#response_check" do

      context "when it receives a success value" do
        subject { output.response_check(string_event, { "response" => "success" } ) }
        it "should return true" do
          expect(subject).to eq(true)
        end
      end

      context "when it receives a non-success value" do
        subject { output.response_check(string_event, { "response" => "fail" } ) }
        it "should return false" do
          expect(subject).to eq(false)
        end
      end

      context "when it receives a non-hash value" do
        subject { output.response_check(string_event, "foo" ) }
        it "should return false" do
          expect(subject).to eq(false)
        end
      end

    end # "#response_check"

    describe "#info_check" do

      context "when it receives a success value" do
        let(:success_msg) { {"response" => "success", "info" => "processed 1; Failed 0; Total 1; seconds spent: 0.000018"} }
        subject { output.info_check(string_event, success_msg) }
        it "should return true" do
          expect(subject).to eq(true)
        end
      end

      context "when it receives a non-success value" do
        let(:fail_msg) { {"response" => "success", "info" => "processed 0; Failed 1; Total 1; seconds spent: 0.000018"} }
        subject { output.info_check(string_event, fail_msg) }
        it "should return false" do
          expect(subject).to eq(false)
        end
      end

      context "when it receives an atypical value" do
        subject { output.info_check(string_event, "foo") }
        it "should return false" do
          expect(subject).to eq(false)
        end
      end

    end # "#info_check"

  end # "Unit Tests"

  describe "Integration Tests", :integration => true do

    before(:all) do
      @zabbix = Longshoreman.new(
        "#{IMAGE}:#{TAG}",    # Image to use
        NAME,                 # Container name
        {
          "Cmd" => ["run"],
          "Tty" => true,
        }                     # Extra options, if any
      )
      zabbix_server_up?(@zabbix.ip, @zabbix.container.rport(ZABBIX_PORT))
    end

    after(:all) do
      @zabbix.cleanup
    end

    describe "#tcp_send", :integration => true do
      let(:port) { @zabbix.container.rport(ZABBIX_PORT) }
      let(:host) { @zabbix.ip }

      context "when the Zabbix server responds with 'success'" do
        subject { output.tcp_send(string_event) }
        it "should return true" do
          expect(subject).to eq(true)
        end
      end

      context "when the Zabbix server responds that some items were unsuccessful" do
        let(:zabhost) { "something_else.example.com" }
        subject { output.tcp_send(string_event) }
        it "should still return true" do
          expect(subject).to eq(true)
        end
      end

      context "when multiple values are sent" do
        let(:port) { @zabbix.container.rport(ZABBIX_PORT) }
        let(:host) { @zabbix.ip }
        let(:multival_event) { LogStash::Event.new(multi_value) }
        let(:m_output) { LogStash::Outputs::Zabbix.new(mzabout) }

        subject { m_output.tcp_send(multival_event) }
        it "should return true" do
          expect(subject).to eq(true)
        end
      end

    end # "#tcp_send", :integration => true

    describe "#send_to_zabbix", :integration => true do
      let(:port) { @zabbix.container.rport(ZABBIX_PORT) }
      let(:host) { @zabbix.ip }

      context "when an event is sent successfully" do
        subject { output.send_to_zabbix(string_event) }
        it "should return true" do
          expect(subject).to eq(true)
        end
      end

      context "when the Zabbix server cannot connect", :integration => true do
        let(:host) { "172.19.74.123" }
        let(:port) { 12345 }
        subject { output.send_to_zabbix(string_event) }
        it "should timeout and return false" do
          expect(subject).to eq(false)
        end
      end

    end # "#send_to_zabbix", :integration => true

  end # "Integration Tests", :integration => true

end # describe LogStash::Outputs::Zabbix


spec/helpers/zabbix_helper.rb
module ZabbixHelper

  # This is used to ensure the (Docker) Zabbix port is up and running
  def port_open?(ip, port, seconds=1)
    Timeout::timeout(seconds) do
      begin
        TCPSocket.new(ip, port).close
        true
      rescue Errno::ECONNREFUSED, Errno::EHOSTUNREACH
        false
      end
    end
  rescue Timeout::Error
    false
  end

  def port_responding?(ip, port)
    # Try for up to 10 seconds to get a response
    10.times do
      if port_open?(ip, port)
        return true
      else
        sleep 1
      end
    end
    false
  end

  def zabbix_server_up?(zabbix_ip, port)
    data = {
        "request" => "sender data",
        "data" => [{
          "host"  => "zabbix.example.com",
          "key"   => "zabbix.key",
          "value" => "This is a log entry.",
          "clock" => 1429100000,
        }],
        "clock" => Time.now.to_i,
    }
    if port_responding?(zabbix_ip, port)
      ###
      ### This is a hacky way to guarantee that Zabbix is responsive, because the
      ### port check alone is insufficient.  TCP tests say the port is open, but
      ### it can take another 2 to 5 seconds (depending on the machine) before it
      ### is responding in the way we need for these tests.
      ###
      resp = ""
      # Try for up to 10 seconds to get a response.
      10.times do
        TCPSocket.open(zabbix_ip, port) do |sock|
          sock.print ZabbixProtocol.dump(data)
          resp = sock.read
        end
        if resp.length == 0
          sleep 1
        else
          return true
        end
      end
      if resp.length == 0
        puts "Zabbix server or db is unreachable"
      end
    else
      puts "Unable to reach Zabbix server on #{zabbix_ip}:#{port}"
    end
    false
  end
  
end


lib/logstash/outputs/zabbix.rb
# encoding: utf-8
require "logstash/outputs/base"
require "logstash/namespace"
require "socket"
require "timeout"
require "zabbix_protocol"

# The Zabbix output is used to send item data (key/value pairs) to a Zabbix
# server.  The event `@timestamp` will automatically be associated with the
# Zabbix item data.
#
# The Zabbix Sender protocol is described at
# https://www.zabbix.org/wiki/Docs/protocols/zabbix_sender/2.0
# Zabbix uses a kind of nested key/value store.
#
# [source,txt]
#     host
#       ├── item1
#       │     └── value1
#       ├── item2
#       │     └── value2
#       ├── ...
#       │     └── ...
#       ├── item_n
#       │     └── value_n
#
# Each "host" is an identifier, and each item is associated with that host.
# Items are typed on the Zabbix side.  You can send numbers as strings and
# Zabbix will Do The Right Thing.
#
# In the Zabbix UI, ensure that your hostname matches the value referenced by
# `zabbix_host`. Create the item with the key as it appears in the field
# referenced by `zabbix_key`.  In the item configuration window, ensure that the
# type dropdown is set to Zabbix Trapper. Also be sure to set the type of
# information that Zabbix should expect for this item.
#
# This plugin does not currently send in batches.  While it is possible to do
# so, this is not supported.  Be careful not to flood your Zabbix server with
# too many events per second.
#
# NOTE: This plugin will log a warning if a necessary field is missing. It will
# not attempt to resend if Zabbix is down, but will log an error message.

class LogStash::Outputs::Zabbix < LogStash::Outputs::Base
  config_name "zabbix"

  # The IP or resolvable hostname where the Zabbix server is running
  config :zabbix_server_host, :validate => :string, :default => "localhost"

  # The port on which the Zabbix server is running
  config :zabbix_server_port, :validate => :number, :default => 10051

  # The field name which holds the Zabbix host name. This can be a sub-field of
  # the @metadata field.
  config :zabbix_host, :validate => :string, :required => true

  # A single field name which holds the value you intend to use as the Zabbix
  # item key. This can be a sub-field of the @metadata field.
  # This directive will be ignored if using `multi_value`
  #
  # IMPORTANT: `zabbix_key` is required if not using `multi_value`.
  #
  config :zabbix_key, :validate => :string

  # The field name which holds the value you want to send.
  # This directive will be ignored if using `multi_value`
  config :zabbix_value, :validate => :string, :default => "message"

  # Use the `multi_value` directive to send multiple key/value pairs.
  # This can be thought of as an array, like:
  #
  # `[ zabbix_key1, zabbix_value1, zabbix_key2, zabbix_value2, ... zabbix_keyN, zabbix_valueN ]`
  #
  # ...where `zabbix_key1` is an instance of `zabbix_key`, and `zabbix_value1`
  # is an instance of `zabbix_value`.  If the field referenced by any
  # `zabbix_key` or `zabbix_value` does not exist, that entry will be ignored.
  #
  # This directive cannot be used in conjunction with the single-value directives
  # `zabbix_key` and `zabbix_value`.
  config :multi_value, :validate => :array

  # The number of seconds to wait before giving up on a connection to the Zabbix
  # server. This number should be very small, otherwise delays in delivery of
  # other outputs could result.
  config :timeout, :validate => :number, :default => 1

  public
  def register
    if !@zabbix_key.nil? && !@multi_value.nil?
      @logger.warn("Cannot use multi_value in conjunction with zabbix_key/zabbix_value.  Ignoring zabbix_key.")
    end

    # We're only going to use @multi_value in the end, so let's build it from
    # @zabbix_key and @zabbix_value if it is empty (single value configuration).
    if @multi_value.nil?
      @multi_value = [ @zabbix_key, @zabbix_value ]
    end
    if @multi_value.length % 2 == 1
      raise LogStash::ConfigurationError, I18n.t("logstash.agent.configuration.invalid_plugin_register",
        :plugin => "output", :type => "zabbix",
        :error => "Invalid zabbix configuration #{@multi_value}. multi_value requires an even number of elements as ['zabbix_key1', 'zabbix_value1', 'zabbix_key2', 'zabbix_value2']")
    end
  end # def register

  public
  def field_check(event, fieldname)
    if !event.get(fieldname)
      @logger.warn("Field referenced by #{fieldname} is missing")
      false
    else
      true
    end
  end

  public
  def kv_check(event, key_field, value_field)
    errors = 0
    for field in [key_field, value_field]
      errors += 1 unless field_check(event, field)
    end
    errors < 1 ? true : false
  end # kv_check

  public
  def validate_fields(event)
    found = []
    (0..@multi_value.length-1).step(2) do |idx|
      if kv_check(event, @multi_value[idx], @multi_value[idx+1])
        found << @multi_value[idx]
        found << @multi_value[idx+1]
      end
    end
    found
  end # validate_fields

  public
  def format_request(event)
    # The nested `clock` value is the event timestamp
    # The ending `clock` value is "now" so Zabbix knows it's not receiving stale
    # data.
    validated = validate_fields(event)
    data = []
    (0..validated.length-1).step(2) do |idx|
      data << {
        "host"  => event.get(@zabbix_host),
        "key"   => event.get(validated[idx]),
        "value" => event.get(validated[idx+1]),
        "clock" => event.get("@timestamp").to_i
      }
    end
    {
      "request" => "sender data",
      "data" => data,
      "clock" => Time.now.to_i,
    }
  end

  def response_check(event, data)
    # {"response"=>"success", "info"=>"Processed 0; Failed 1; Total 1; seconds spent: 0.000018"}
    unless data["response"] == "success"
      @logger.error("Failed to send event to Zabbix",
        :zabbix_response => data,
        :event => event
      )
      false
    else
      true
    end
  end

  def info_check(event, data)
    # {"response"=>"success", "info"=>"processed 0; Failed 1; Total 1; seconds spent: 0.000018"}
    if !data.is_a?(Hash)
      @logger.error("Zabbix server at #{@zabbix_server_host} responded atypically.",
        :returned_data => data
      )
      return false
    end
    # Prune the semicolons, then turn it into an array
    info = data["info"].tr(';', '').split()
    # ["processed", "0", "Failed", "1", "Total", "1", "seconds", "spent:", "0.000018"]
    failed = info[3].to_i
    total = info[5].to_i
    if failed == total
      @logger.warn("Zabbix server at #{@zabbix_server_host} rejected all items sent.",
        :zabbix_host => event.get(@zabbix_host)
      )
      false
    elsif failed > 0
      @logger.warn("Zabbix server at #{@zabbix_server_host} rejected #{info[3]} item(s).",
        :zabbix_host => event.get(@zabbix_host)
      )
      false
    elsif failed == 0 && total > 0
      true
    else
      false
    end
  end

  def tcp_send(event)
    begin
      TCPSocket.open(@zabbix_server_host, @zabbix_server_port) do |sock|
        data = format_request(event)
        sock.print ZabbixProtocol.dump(data)
        resp = ZabbixProtocol.load(sock.read)
        @logger.debug? and @logger.debug("Zabbix server response",
                              :response => resp, :data => data)
        # Log whether the key/value pairs accepted
        info_check(event, resp)
        # Did the message get received by Zabbix?
        response_check(event, resp)
      end
    rescue Errno::ECONNREFUSED, Errno::EHOSTUNREACH, Errno::ECONNRESET
      @logger.error("Connection error.  Unable to connect to Zabbix server",
        :server => @zabbix_server_host,
        :port => @zabbix_server_port.to_s
      )
      false
    end
  end

  def send_to_zabbix(event)
    begin
      Timeout::timeout(@timeout) do
        tcp_send(event)
      end
    rescue Timeout::Error
      @logger.warn("Connection attempt to Zabbix server timed out.",
        :server => @zabbix_server_host,
        :port => @zabbix_server_port.to_s,
        :timeout => @timeout.to_s
      )
      false
    end
  end

  public
  def receive(event)

    return unless field_check(event, @zabbix_host)
    send_to_zabbix(event)
  end # def event

end # class LogStash::Outputs::Zabbix
